web3/middleware/signing.py

```diff
@@ -0,0 +1,78 @@
+from eth_keys.datatypes import PrivateKey
+from eth_utils import is_same_address
+
+from web3.utils.datastructures import HexBytes
+from web3.exceptions import InvalidAddress
+# from web3.utils.signing import LocalAccount
+
+import logging
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def construct_transaction_signing_middleware(private_key):
+    """Intercept a transaction in flight and sign it.
+
+    Args:
+        private_key (TYPE): Private key used to sign the transaction.
+                            Accepts the following formats:
+                            - `PrivateKey` object from `eth-keys`
+                            - `LocalAccount` object from `web3.eth.account`
+                            - raw bytes
+                            - TODO: hex encoded string
+
+
+    Returns:
+        TYPE: Description
+    """
+
+    # Check if the private_key is either
+    # - eth_keys.keys.PrivateKey
+    # - web3.eth.account.create().privateKey -- can just check that its HexBytes
+    # - bytes
+    # TODO: - hexbytes
+    valid_key_types = (PrivateKey, HexBytes, bytes)
+
+    _private_key = private_key if isinstance(private_key, valid_key_types) else None
+    if not _private_key:
+        raise ValueError('Private Key is invalid.')
+
+    def transaction_signing_middleware(make_request, web3):
+        def middleware(method, params):
+            # Only operate on the `eth.sendTransaction` method
+            # Only operate on if the private key matches the public key in the transaction
+            # Note: params == transaction in this case.
+            if method != 'eth_sendTransaction':
+                return make_request(method, params)
+
+            transaction = params
```

Comment:
I don't believe this is correct.  `params` should be a `list` with a single element which is the transaction dictionary.  I think this should be changed to `transaction = params[0]` (though it would be good to do a pre-check to ensure that `params` and `params[0]` are both well formed:

- list with a single element
- single element is a dictionary